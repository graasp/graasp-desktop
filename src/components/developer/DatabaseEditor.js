import React, { Component } from 'react';
import _ from 'lodash';
import { connect } from 'react-redux';
import ReactJson from 'react-json-view';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import { withTranslation } from 'react-i18next';
import { withStyles } from '@material-ui/core';
import { getDatabase, setDatabase } from '../../actions';
import Loader from '../common/Loader';
import Styles from '../../Styles';
import SampleDatabase from '../../data/sample.json';

export class DatabaseEditor extends Component {
  static propTypes = {
    t: PropTypes.func.isRequired,
    classes: PropTypes.shape({
      button: PropTypes.string.isRequired,
    }).isRequired,
    dispatchGetDatabase: PropTypes.func.isRequired,
    dispatchSetDatabase: PropTypes.func.isRequired,
    database: PropTypes.shape({
      user: PropTypes.object,
      spaces: PropTypes.array,
    }),
  };

  static defaultProps = {
    database: {},
  };

  componentDidMount() {
    const { dispatchGetDatabase } = this.props;
    dispatchGetDatabase();
  }

  handleEdit = ({ updated_src: updatedSrc }) => {
    const { dispatchSetDatabase } = this.props;
    dispatchSetDatabase(updatedSrc);
  };

  handleUseSampleDatabase = () => {
    const { dispatchSetDatabase } = this.props;
    dispatchSetDatabase(SampleDatabase);
  };

  render() {
    const { database, t, classes } = this.props;

    if (!database || _.isEmpty(database)) {
      return <Loader />;
    }

    return (
      <div>
        <Typography variant="h6">{t('Manually Edit the Database')}</Typography>
        <ReactJson
          collapsed
          src={database}
          onEdit={this.handleEdit}
          onAdd={this.handleEdit}
          onDelete={this.handleEdit}
        />
        <br />
        <Button
          variant="contained"
          className={classes.button}
          onClick={this.handleUseSampleDatabase}
          color="primary"
        >
          {t('Use Sample Database')}
        </Button>
      </div>
    );
  }
}

const mapStateToProps = ({ Developer }) => ({
  database: Developer.get('database'),
});

const mapDispatchToProps = {
  dispatchGetDatabase: getDatabase,
  dispatchSetDatabase: setDatabase,
};

const StyledComponent = withStyles(Styles, { withTheme: true })(DatabaseEditor);
const TranslatedComponent = withTranslation()(StyledComponent);
const ConnectedComponent = connect(
  mapStateToProps,
  mapDispatchToProps
)(TranslatedComponent);

export default ConnectedComponent;
