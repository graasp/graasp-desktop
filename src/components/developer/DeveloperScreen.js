import React, { Component } from 'react';
import classNames from 'classnames';
import { withStyles } from '@material-ui/core';
import { withTranslation } from 'react-i18next';
import { withRouter } from 'react-router';
import Typography from '@material-ui/core/Typography';
import CssBaseline from '@material-ui/core/CssBaseline';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import IconButton from '@material-ui/core/IconButton';
import MenuIcon from '@material-ui/icons/Menu';
import ChevronLeftIcon from '@material-ui/icons/ChevronLeft';
import ChevronRightIcon from '@material-ui/icons/ChevronRight';
import Drawer from '@material-ui/core/Drawer';
import Divider from '@material-ui/core/Divider';
import PropTypes from 'prop-types';
import MainMenu from '../common/MainMenu';
import DatabaseEditor from './DatabaseEditor';
import Styles from '../../Styles';
import Banner from '../common/Banner';

class DeveloperScreen extends Component {
  state = {
    open: false,
  };

  static propTypes = {
    t: PropTypes.func.isRequired,
    classes: PropTypes.shape({}).isRequired,
    theme: PropTypes.shape({}).isRequired,
    history: PropTypes.shape({
      replace: PropTypes.func.isRequired,
    }).isRequired,
    i18n: PropTypes.shape({
      changeLanguage: PropTypes.func.isRequired,
    }).isRequired,
  };

  handleDrawerOpen = () => {
    this.setState({ open: true });
  };

  handleDrawerClose = () => {
    this.setState({ open: false });
  };

  render() {
    const { classes, theme, t } = this.props;
    const { open } = this.state;

    return (
      <div className={classes.root}>
        <CssBaseline />
        <AppBar
          position="fixed"
          className={classNames(classes.appBar, {
            [classes.appBarShift]: open,
          })}
        >
          <Toolbar disableGutters={!open}>
            <IconButton
              color="inherit"
              aria-label="Open drawer"
              onClick={this.handleDrawerOpen}
              className={classNames(classes.menuButton, open && classes.hide)}
            >
              <MenuIcon />
            </IconButton>
          </Toolbar>
        </AppBar>
        <Drawer
          className={classes.drawer}
          variant="persistent"
          anchor="left"
          open={open}
          classes={{
            paper: classes.drawerPaper,
          }}
        >
          <div className={classes.drawerHeader}>
            <IconButton onClick={this.handleDrawerClose}>
              {theme.direction === 'ltr' ? (
                <ChevronLeftIcon />
              ) : (
                <ChevronRightIcon />
              )}
            </IconButton>
          </div>
          <Divider />
          <MainMenu />
        </Drawer>
        <main
          className={classNames(classes.content, {
            [classes.contentShift]: open,
          })}
        >
          <div className={classes.drawerHeader} />
          <div className={classes.developer}>
            <Typography variant="h4" className={classes.screenTitle}>
              {t('Developer')}
            </Typography>
            <br />
            <Banner
              text={t(
                'Danger Zone! Proceed with caution as changes to this section might lead to data loss.'
              )}
              type="error"
            />
            <DatabaseEditor />
          </div>
        </main>
      </div>
    );
  }
}

const StyledComponent = withStyles(Styles, { withTheme: true })(
  DeveloperScreen
);

const TranslatedComponent = withTranslation()(StyledComponent);

export default withRouter(TranslatedComponent);
