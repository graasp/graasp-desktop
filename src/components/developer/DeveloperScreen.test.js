import React from 'react';
import { shallow } from 'enzyme';
import Typography from '@material-ui/core/Typography';
import { DeveloperScreen } from './DeveloperScreen';
import DatabaseEditor from './DatabaseEditor';
import Banner from '../common/Banner';

const createDeveloperScreenProps = () => {
  return {
    t: text => text,
    classes: {
      root: '',
      appBar: '',
      appBarShift: '',
      menuButton: '',
      hide: '',
      drawer: '',
      drawerPaper: '',
      drawerHeader: '',
      content: '',
      contentShift: '',
      developer: 'developer',
      screenTitle: 'screenTitle',
    },
    theme: {
      direction: 'ltr',
    },
    history: {
      replace: jest.fn(),
    },
    i18n: {
      changeLanguage: jest.fn(),
    },
  };
};

describe('<DeveloperScreen />', () => {
  let wrapper;

  beforeAll(() => {
    const props = createDeveloperScreenProps();
    // eslint-disable-next-line react/jsx-props-no-spreading
    wrapper = shallow(<DeveloperScreen {...props} />);
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders one <Typography /> component with text "Developer"', () => {
    const typography = wrapper.find(Typography);
    expect(typography).toHaveLength(1);
    expect(typography.contains('Developer')).toBeTruthy();
  });

  it('renders one <Banner /> component with error', () => {
    const banner = wrapper.find(Banner).find({ type: 'error' });
    expect(banner).toHaveLength(1);
  });

  it('renders one <DatabaseEditor /> component', () => {
    const databaseEditor = wrapper.find(DatabaseEditor);
    expect(databaseEditor).toHaveLength(1);
  });
});
