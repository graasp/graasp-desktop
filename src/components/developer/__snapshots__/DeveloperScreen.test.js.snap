// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeveloperScreen /> renders correctly 1`] = `ShallowWrapper {}`;
